File: day17/ans.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == '1':
                    # Make itself and all neighbours zero in BFS fashion
                    grid[i][j] = '0'
                    
                    q = deque()
                    q.append((i,j))
                    while q:
                        r, c = q.popleft()
                        
                        if r != len(grid) - 1 and grid[r+1][c] == '1':
                            grid[r+1][c] = '0'
                            q.append((r+1, c))
                        
                        if r != 0 and grid[r-1][c] == '1':
                            grid[r-1][c] = '0'
                            q.append((r-1, c))

                        if c != len(grid[r]) - 1 and grid[r][c+1] == '1':
                            grid[r][c+1] = '0'
                            q.append((r, c+1))
                        
                        if c != 0 and grid[r][c - 1] == '1':
                            grid[r][c-1] = '0'
                            q.append((r, c-1))
                    
                    #increment island count
                    count += 1
        
        return count    

File: day17/ans.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        int count = 0;
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[i].size(); j++) {
                if(grid[i][j] == '1') {
                    //cout << "I am Here!!\n";
                    count += 1;
                    makeNeighboursZero(grid, i , j);
                }
            }
        }
        
        return count;
    }
    
    void makeNeighboursZero(vector<vector<char>>& grid, int i, int j) {
        //Base condition is implicit when all the neighbours are zero
        
        if(i != grid.size() - 1 && grid[i+1][j] == '1') {
            grid[i+1][j] = '0';
            makeNeighboursZero(grid, i + 1, j);
        }
        
        if(i != 0 && grid[i - 1][j] == '1') {
            grid[i - 1][j] = '0';
            makeNeighboursZero(grid, i - 1, j);
        }
        
        if(j != grid[i].size() - 1 && grid[i][j + 1] == '1') {
            grid[i][j + 1] = '0';
            makeNeighboursZero(grid, i, j + 1);
        }
        
        if(j != 0 && grid[i][j - 1] == '1') {
            grid[i][j - 1] = '0';
            makeNeighboursZero(grid, i, j - 1);
        }
        
        return;
    }
};
